### PR TITLE
RFC: explicitly set strict flag in `zip` calls (enforce `B905`), and ignore `UP038`

### LINKED ISSUE #4749: MNT: how should we handle Python 3.10 becoming our oldest supported version ?
Python 3.10 is next in line to become our oldest supported version (currently it's 3.9), sometime about October 2024 (extrapolating from the last couple year's pace).
This version comes with a couple changes that will be picked up but our linters (`pyugrade` and `flake8-bugbear` via `ruff`) and that would affect a large number of lines, so I'd like to start the discussion early to decide what to do about them. 

Here are the rules I know about

## [pyupgrade UP038](https://docs.astral.sh/ruff/rules/non-pep604-isinstance/)

Basically this would change
```python
isinstance(x, (int, float))
```
to
```python
isinstance(x, int | float)
```
It's noted that the more modern ("good") alternative is actually slower, so we might want to just ignore this specific rule. 

## [flake8-bugbear B905](https://docs.astral.sh/ruff/rules/zip-without-explicit-strict/)
Python 3.10 added a `strict` argument to the builtin `zip` function which is `False` by default for backward compatibility.
This rule checks that `strict` is explicitly set everywhere, which is always safer or more explicit than leaving it unset.
I don't think it'd be controversial to follow this rule ? It'd a bit tedious to retro-actively set it everywhere but I'd happily volunteer for this boring task. It can actually be addressed right now by using `yt._maintenance.backports.zip` which imitate the Python 3.10 function.

### PR DESCRIPTION
**Pull request by neutrinoceros**
Close #4749

This is a mostly automated style change, but I'm using the test suite to detect places where `strict=False` is either the intended behavior or a bug.

I'll also carefully review this myself to improve coverage.

### CODE REVIEW

File: yt/data_objects/profiles.py
Comment (neutrinoceros) [inline yt/data_objects/profiles.py:1476]: This is one very case where it's apparent that `strict=False` is the intended behavior (found thanks to a test).

File: yt/data_objects/analyzer_objects.py
Comment (neutrinoceros) [inline yt/data_objects/analyzer_objects.py:17]: from the context, it seems that `strict=False` is clearly intended here

File: yt/data_objects/derived_quantities.py
Comment (neutrinoceros) [inline yt/data_objects/derived_quantities.py:633]: This is only safe if `len(values)%2==0` is guaranteed. If not, `strict=False` matches the existing behavior but is possibly buggy (one value is silently dropped)

**Comment by neutrinoceros** [comment]
On second though this isn't a pure stylistic choice, as it *may* (however unlikely) have undesirable effects. Re-triaging as a refactor.

File: yt/frontends/stream/data_structures.py
Comment (neutrinoceros) [inline yt/frontends/stream/data_structures.py:431]: this is a good example of why enforcing `strict=True` helps finding bugs

File: yt/data_objects/profiles.py
Comment (neutrinoceros) [inline yt/data_objects/profiles.py:1324]: highlighting this bit as a bugfix

File: yt/frontends/athena/io.py
Comment (neutrinoceros) [inline yt/frontends/athena/io.py:29]: this is a case where the original code just *looks* wrong, but it's been there for 12 years so either it's not really a bug (but it should have a comment), or it's actually not used by anyone...

File: yt/utilities/fortran_utils.py
Comment (neutrinoceros) [inline yt/utilities/fortran_utils.py:78]: not strictly backward compatible if we assume anyone relies on the exact exception type here, but I'm taking a bet that @cphyc doesn't mind.

**Review by neutrinoceros — COMMENTED** [review]
Finishing my self-review with a couple notes for reviewers:
First of, if you're wondering why I write
```python
if sys.version_info >= (3, 10):
    pass
else:
    from yt._maintenance.backports import zip
```
instead of the simpler
```python
if sys.version_info < (3, 10):
    from yt._maintenance.backports import zip
```
It is because ruff ([rule `UP036`](https://docs.astral.sh/ruff/rules/outdated-version-block/)) can automatically cleanup the former but not the latter when we switch `python-requires` to `>=3.10`. So it's more verbose now but easier to cleanup in the future.

2) in a couple places, I actively changed the code around a `zip` to fix apparent bugs. These should be highlighted by my own comments.

3) Please pay special attention in cases where I set `strict=False`: in most cases, I saw no actual test failing with `strict=True` but the *context* around `zip` made me think `strict=False` was the intended behavior. I may be wrong.

4) Some `zip` calls were zipping some small literal lists, and I refactored them to as lists of tuples to avoid `zip` entirely, which I found was more readable. Maybe I made a mistake in a case or two, so please pay extra attention there too.

**Comment by neutrinoceros** [comment]
I know this isn't exactly an exciting read, but can I ask @matthewturk, @chrishavlin or @cphyc for a review ?

**Comment by chrishavlin** [comment]
Yes! I actually started it a few days ago but didn't finish up. Can definitely get to it in the next few days.

File: yt/data_objects/data_containers.py
Comment (matthewturk) [inline yt/data_objects/data_containers.py:804]: If an error were to occur, I think this is a candidate location.

File: yt/data_objects/data_containers.py
Comment (matthewturk) [inline yt/data_objects/data_containers.py:856]: If an error were to occur, I think this is a candidate location.

File: yt/data_objects/derived_quantities.py
Comment (matthewturk) [reply yt/data_objects/derived_quantities.py:631]: I think using strict here is the right move, as a check on argument inputs.

File: yt/data_objects/profiles.py
Comment (matthewturk) [reply yt/data_objects/profiles.py:1322]: Thank you.  I have an inkling somewhere in my head that there's an extant issue this could be related to but searching does not prove me correct.

File: yt/data_objects/profiles.py
Comment (matthewturk) [reply yt/data_objects/profiles.py:1476]: Great catch.  Yes, I think this is assuming that it'll run out of axes based on the dimensionality.

File: yt/fields/particle_fields.py
Comment (matthewturk) [inline yt/fields/particle_fields.py:200]: this is indeed an improvement

File: yt/frontends/art/io.py
Comment (matthewturk) [inline yt/frontends/art/io.py:132]: If an error were to occur, I think this is a candidate location.

File: yt/frontends/artio/definitions.py
Comment (matthewturk) [inline yt/frontends/artio/definitions.py:36]: This is correct behavior -- but it makes me realize that we are using a dict for a 1:1 mapping, which is not necessarily true.  Fortunately it is here!

File: yt/frontends/athena/io.py
Comment (matthewturk) [reply yt/frontends/athena/io.py:27]: I don't see any references to it -- this may be vestigial.

File: yt/frontends/enzo_e/data_structures.py
Comment (matthewturk) [inline yt/frontends/enzo_e/data_structures.py:478]: I can't help but feel like I probably initiated this idiom that you're undoing in several places ... and I think undoing it is the right thing!

File: yt/frontends/enzo_e/misc.py
Comment (matthewturk) [inline yt/frontends/enzo_e/misc.py:97]: if this errors it will be a very valuable error check

File: yt/frontends/tipsy/data_structures.py
Comment (matthewturk) [inline yt/frontends/tipsy/data_structures.py:129]: This one I'm less certain of, in that it may be the case that we don't want to force the specification of all fields and just let some go.  But, that's not necessarily what we should build in to doing this check, so :+1: .

File: yt/utilities/answer_testing/framework.py
Comment (matthewturk) [inline yt/utilities/answer_testing/framework.py:776]: this is a good additional check

Comment (matthewturk) [reply yt/utilities/answer_testing/framework.py:776]: although, it may result in slightly less informative errors if the lengths are different

File: yt/utilities/fortran_utils.py
Comment (matthewturk) [inline yt/utilities/fortran_utils.py:135]: I do think the other version was somewhat better, as otherwise we're left deducing what caused the iterators to be of different lengths.

**Review by matthewturk — APPROVED** [review]
I left some comments where I think there *might* be issues, but I want to note that I think there are no problems with any of this, and any bugs it finds would be an improvement.  Thanks for taking the time and leaving such helpful comments on it.

File: yt/frontends/athena/io.py
Comment (neutrinoceros) [reply yt/frontends/athena/io.py:27]: you're right. And it's private anyway so let me just remove it now.

File: yt/utilities/answer_testing/framework.py
Comment (neutrinoceros) [reply yt/utilities/answer_testing/framework.py:776]: not sure what you mean here: if lengths are different, there might be no error at all (without strict=True)

Comment (matthewturk) [reply yt/utilities/answer_testing/framework.py:776]: Oh, I just mean, that the old way wouldn't pick that up *as-is* but the new way will show a weird error.  It's NBD.

File: yt/utilities/fortran_utils.py
Comment (neutrinoceros) [reply yt/utilities/fortran_utils.py:135]: FTR I'm waiting on @chrishavlin to weigh in but happy to change it if he agrees

**Review by chrishavlin — APPROVED** [review]
Generally looks good! some minor questions/remarks so i'll hit approve now (assume approval if it gets dismissed!)

I'll also say I had a hard time evaluating the changes in the code for some of the frontends (particularly where the existing zips related to file structures....), but I think we have to trust (hope?) that the tests would catch errors there.

File: yt/utilities/fortran_utils.py
Comment (chrishavlin) [reply yt/utilities/fortran_utils.py:135]: i think i prefer keeping the OSError, but don't feel too strongly about it.

File: yt/data_objects/construction_data_containers.py
Comment (neutrinoceros) [reply yt/data_objects/construction_data_containers.py:158]: Good point. One could argue that in the presence of `strict=True` `pids` does play some role in terms of validation but it's convoluted and beyond the point. I imagine `pids` just came from copy-pasta. Let me clean this up now.

Comment (neutrinoceros) [reply yt/data_objects/construction_data_containers.py:158]: wait no, `i` (element from `pids`) *is* used in this loop. Not touching this here after all. (but I take your point that it is confusing !)

File: yt/data_objects/region_expression.py
Comment (neutrinoceros) [reply yt/data_objects/region_expression.py:173]: nice catch, let me try this

File: yt/frontends/athena/io.py
Comment (neutrinoceros) [reply yt/frontends/athena/io.py]: thanks for catching this ! clearly vestigial from a auto-fixing

File: yt/utilities/fortran_utils.py
Comment (neutrinoceros) [reply yt/utilities/fortran_utils.py:143]: reverted !

**Comment by neutrinoceros** [comment]
Can I get another approval ? Seems that I dismissed the ones I got when I pushed another clean up commit

**Review by chrishavlin — APPROVED** [review]

**Comment by neutrinoceros** [comment]
Let's get this in before it starts rotting

**Pull request merged**
